### PR TITLE
[JENKINS-55423] Support custom Jenkins Root URLs for webhook setup

### PR DESCRIPTION
**Pull request by jimklimov**
https://issues.jenkins-ci.org/browse/JENKINS-55423

In some cases, the Jenkins server can be seen to its clients (here the Bitbucket Cloud or Server instance) under a different URL than it declares in its local networking segment. This PR adds an optional configurable setting for each endpoint to use a custom Jenkins Root URL when managing webhooks for that endpoint. Tests are extended accordingly, which was a large part of the effort in the end.

Note that while each BitbucketSCMSource object has its copy of the server and root URL strings, and can in theory be instantiated alone, its lookup of a BitbucketEndpointConfiguration::readResolveServerUrl() creates a dummy global config entry for the serverUrl if it was missing (and this dummy has all fields defaulted to nothing). Due to this, tests about this feature have to preload (add or update) the endpoint config into the global list first, and then run tests against it.

### CODE REVIEW

**Comment by jimklimov** [comment]
CC @casz @twz123

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMNavigator.java [hunk 1]
```diff
@@ -145,6 +145,14 @@
     @Restricted(NoExternalUse.class)
     @RestrictedSince("2.2.0")
     private transient String bitbucketServerUrl;
+    @Deprecated
+    @Restricted(NoExternalUse.class)
+    @RestrictedSince("2.2.0")
+    private transient String bitbucketJenkinsRootUrl;
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMNavigator.java]: Why two transient values? and why are they both marked deprecated?

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMNavigator.java [hunk 2]
```diff
@@ -180,6 +188,13 @@ private Object readResolve() throws ObjectStreamException {
         if (serverUrl == null) {
             serverUrl = BitbucketEndpointConfiguration.get().readResolveServerUrl(bitbucketServerUrl);
         }
+        if (serverUrl == null) {
+            LOGGER.log(Level.WARNING, "BitbucketSCMNavigator::readResolve : serverUrl is still empty");
+        } else {
+            LOGGER.log(Level.FINEST, "BitbucketSCMNavigator::readResolve : current bitbucketJenkinsRootUrl={0}", bitbucketJenkinsRootUrl == null ? "<null>" : "'" + bitbucketJenkinsRootUrl + "'" );
+            this.updateBitbucketJenkinsRootUrl(serverUrl);
+            LOGGER.log(Level.FINEST, "BitbucketSCMNavigator::readResolve : updated bitbucketJenkinsRootUrl={0}", bitbucketJenkinsRootUrl == null ? "<null>" : "'" + bitbucketJenkinsRootUrl + "'" );
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMNavigator.java]: All this logging feels redundant

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMNavigator.java [hunk 3]
```diff
@@ -210,6 +225,81 @@ private Object readResolve() throws ObjectStreamException {
         return this;
     }
 
+    /*
+     * Optionally update bitbucketJenkinsRootUrl if it was not set
+     * before (which causes normalization of the URL string value),
+     * or if the cached endpointcfgJenkinsRootUrl changed. Also updates
+     * this cached value as needed (was not set or did change).
+     */
+    @SuppressFBWarnings(value = "RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE",
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMNavigator.java]: This seems like code smell

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMNavigator.java [hunk 4]
```diff
@@ -359,6 +449,52 @@ public String getBitbucketServerUrl() {
         return serverUrl;
     }
 
+    @Deprecated
+    @Restricted(DoNotUse.class)
+    @RestrictedSince("2.2.0")
+    @DataBoundSetter
+    public void setBitbucketJenkinsRootUrl(String rootUrl) {
+        if (rootUrl == null || rootUrl.equals("")) {
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMNavigator.java]: use util method to `Util.fixTrimAndEmpty`

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSource.java [hunk 5]
```diff
@@ -306,6 +337,81 @@ private Object readResolve() throws ObjectStreamException {
         return this;
     }
 
+    /*
+     * Optionally update bitbucketJenkinsRootUrl if it was not set
+     * before (which causes normalization of the URL string value),
+     * or if the cached endpointcfgJenkinsRootUrl changed. Also updates
+     * this cached value as needed (was not set or did change).
+     */
+    @SuppressFBWarnings(value = "RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE",
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSource.java]: This code smells double up, why all this duplicated code?

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/AbstractBitbucketEndpoint.java [hunk 6]
```diff
@@ -55,16 +66,42 @@
...
      * Constructor.
      *
      * @param manageHooks   {@code true} if and only if Jenkins is supposed to auto-manage hooks for this end-point.
      * @param credentialsId The {@link StandardCredentials#getId()} of the credentials to use for
      *                      auto-management of hooks.
+     * @param bitbucketJenkinsRootUrl The custom (or empty for global setting) Jenkins Root URL
+     *                      auto-management of hooks.
      */
-    AbstractBitbucketEndpoint(boolean manageHooks, @CheckForNull String credentialsId) {
+    AbstractBitbucketEndpoint(boolean manageHooks, @CheckForNull String credentialsId, @CheckForNull String bitbucketJenkinsRootUrl) {
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/AbstractBitbucketEndpoint.java]: You are passing the objects in the wrong order. 
The abstractBitbucketEndpoint should own this Jenkins URL might differ per endpoint.
It should just call `this.getJenkinsRootUrl()` which is a method that either takes the custom Jenkins URL or the instance root URL. I would just skip the whole transient normalized thingy... and choose to store it on the bitbucketJenkinsRootUrl.

**Comment by jimklimov** [comment]
Thanks for the thorough review, I agree that the coding could be better if I knew and understood Java better :)

Looking to analyze how I got this far, I mostly went along the codebase where similar manipulations (e.g. with `serverUrl`) were being done. And that's probably how all the duplication appeared - the `(bitbucket)serverUrl` variables have a copy in the BitbucketSCMSource and BitbucketSCMNavigator, for whatever reason, though maybe it was just for requesting into the currently defined endpoint with same URL value. I do not (or did not) understand those patterns really well. Another cause was the self-testing, since the custom root URLs had to land somewhere in the class to be loaded from XMLs, though this was a consideration from a point before I managed to get a working `loadBEC()`... and it took several takes to get made.

Likewise for superfluous logging, since the entanglement of these variables (and interactions like that requesting an endpoint config creates an empty-fielded dummy entry if the `serverUrl` was not known before), making sense of what happens to those URLs on a production server was complicated.

At the very least, currently the code both compiles, passes tests (and has custom tests for new behavior), and does what I wanted of it in production :) This is a good starting point to rectify the programming aspects, probably starting with eviction of the rooturl variables from classes other than the abstract endpoint.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMNavigator.java (hunk 1 above)
Comment (jimklimov) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMNavigator.java]: Copypasted from other fields that worked. Otherwise, sorry but no idea what this really means (not a Java dev here) :-)

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMNavigator.java (hunk 3 above)
Comment (jimklimov) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMNavigator.java]: At least in the current iteration (or some earlier points in the development intent), there is a difference between a `null` value which means that no setting was applied into the string yet (so it should be re-requested from the endpoint configs) vs. a known empty-string setting (so we have loaded a config, and it did not specify a custom URL).

Either way, when we do not have a definitive custom URL set, we return a (normalized) global Jenkins Root URL. Normalization of the webhook URL compared to what users can enter proved important, e.g. a missing trailing slash causes mayhem. Maybe ensuring of this slash can be moved to where the value is used, but then the code would be duplicated there, and I was not sure I tracked down all the consumers :\

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/AbstractBitbucketEndpoint.java (hunk 6 above)
Comment (jimklimov) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/AbstractBitbucketEndpoint.java]: I am not sure how to interpret this, could you please rephrase in layman terms? :-)
What would the correct order be?

As for the transient normalized, as I commented elsewhere, it is important that the URL string returned to callers conforms to some of their expectations. Normalization is a moderately costly but definitive translation from one string (user's input stored in config) to another string (normalized value), so I figured it made sense to only normalize once and not on every getter access.

**Comment by casz** [comment]
We all have to start somewhere :grimacing: :innocent:
The code looks fine except for being duplicated 😂

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMNavigator.java (hunk 4 above)
Comment (jimklimov) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMNavigator.java]: found a `fixEmptyAndTrim` in Jenkins, seems to be it :)

**Comment by jimklimov** [comment]
I've hopefully addressed as much as I could about code duplication, now the feature change is almost all concentrated in `AbstractBitbucketEndpoint` class.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/AbstractBitbucketEndpoint.java (hunk 6 above)
Comment (casz) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/AbstractBitbucketEndpoint.java]: I don't think this should be passed to the constructor

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/AbstractBitbucketEndpoint.java [hunk 7]
```diff
@@ -55,16 +66,42 @@
...
     private final String credentialsId;
 
+    /**
+     * Jenkins Server Root URL to be used by that Bitbucket endpoint.
+     * The global setting from Jenkins.getActiveInstance().getRootUrl()
+     * will be used if this field is null or equals an empty string.
+     * This variable is bound to the UI, so an empty value is saved
+     * and returned by getter as such.
+     */
+    private final String bitbucketJenkinsRootUrl;
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/AbstractBitbucketEndpoint.java]: Should not be final

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/AbstractBitbucketEndpoint.java [hunk 8]
```diff
@@ -55,16 +66,42 @@
...
         this.manageHooks = manageHooks && StringUtils.isNotBlank(credentialsId);
-        this.credentialsId = manageHooks ? credentialsId : null;
+        if (this.manageHooks) {
+            this.credentialsId = credentialsId;
+            this.bitbucketJenkinsRootUrl = bitbucketJenkinsRootUrl ;
+        } else {
+            this.credentialsId = null;
+            this.bitbucketJenkinsRootUrl = "";
+        }
+        this.setEndpointJenkinsRootUrl(this.bitbucketJenkinsRootUrl);
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/AbstractBitbucketEndpoint.java]: This line could be moved to a readResolve, so when Jenkins reads the XML config it will normalize and set the transient property on restart. And on set of `bitbucketJenkinsRootUrl` it should call `setEndpointJenkinsRootUrl`

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/BitbucketCloudEndpoint.java [hunk 9]
```diff
@@ -77,10 +77,12 @@ public BitbucketCloudEndpoint(boolean manageHooks, @CheckForNull String credenti
...
      * @param credentialsId The {@link StandardCredentials#getId()} of the credentials to use for
      *                      auto-management of hooks.
-     */
+     * @param bitbucketJenkinsRootUrl The custom (or empty for global setting) Jenkins Root URL
+     *                      auto-management of hooks.
+      */
     @DataBoundConstructor
-    public BitbucketCloudEndpoint(boolean enableCache, int teamCacheDuration, int repositoriesCacheDuration, boolean manageHooks, @CheckForNull String credentialsId) {
-        super(manageHooks, credentialsId);
+    public BitbucketCloudEndpoint(boolean enableCache, int teamCacheDuration, int repositoriesCacheDuration, boolean manageHooks, @CheckForNull String credentialsId, @CheckForNull String bitbucketJenkinsRootUrl) {
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/BitbucketCloudEndpoint.java]: I would prefer that this was not added to the `@DataboundConstructor`.

Instead, use a `@DataboundSetter` since the value is not mandatory. This also removes the need for passing empty strings around.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/BitbucketEndpointConfiguration.java [hunk 10]
```diff
@@ -95,9 +95,9 @@ public String readResolveServerUrl(@CheckForNull String bitbucketServerUrl) {
             if (BitbucketCloudEndpoint.SERVER_URL.equals(serverUrl)
                     || BitbucketCloudEndpoint.BAD_SERVER_URL.equals(serverUrl)) {
                 // exception case
-                addEndpoint(new BitbucketCloudEndpoint(false, null));
+                addEndpoint(new BitbucketCloudEndpoint(false, null, ""));
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/BitbucketEndpointConfiguration.java]: The requested changes will also remove this empty string passing 😇

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/BitbucketEndpointConfiguration.java [hunk 11]
```diff
@@ -167,12 +167,12 @@ public synchronized void setEndpoints(@CheckForNull List<? extends AbstractBitbu
             } else if (!(endpoint instanceof BitbucketCloudEndpoint)
                     && BitbucketCloudEndpoint.SERVER_URL.equals(serverUrl)) {
                 // fix type for the special case
-                iterator.set(new BitbucketCloudEndpoint(endpoint.isManageHooks(), endpoint.getCredentialsId()));
+                iterator.set(new BitbucketCloudEndpoint(endpoint.isManageHooks(), endpoint.getCredentialsId(), endpoint.getBitbucketJenkinsRootUrl()));
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/BitbucketEndpointConfiguration.java]: Perhaps for this use case, it makes sense to create a new constructor instead of adding more to the databound constructor.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/BitbucketServerEndpoint.java [hunk 12]
```diff
@@ -91,11 +91,13 @@
      * @param manageHooks   {@code true} if and only if Jenkins is supposed to auto-manage hooks for this end-point.
      * @param credentialsId The {@link StandardCredentials#getId()} of the credentials to use for
      *                      auto-management of hooks.
-     */
+     * @param bitbucketJenkinsRootUrl The custom (or empty for global setting) Jenkins Root URL
+     *                      auto-management of hooks.
+      */
     @DataBoundConstructor
     public BitbucketServerEndpoint(@CheckForNull String displayName, @NonNull String serverUrl, boolean manageHooks,
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/BitbucketServerEndpoint.java]: Same as previously I would prefer it was using a `@DataBoundSetter` for bitbucketJenkinsRootUrl and than create another constructor that is not databound to allow receiving bitbucketJenkinsRootUrl on constructor if it makes coded behavior better.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/AbstractBitbucketEndpoint.java (hunk 8 above)
Comment (casz) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/AbstractBitbucketEndpoint.java]: Don't know if the readResolve can be on the abstract class or it has to be in both endpoints.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMNavigator.java [hunk 13]
```diff
@@ -359,6 +362,18 @@ public String getBitbucketServerUrl() {
         return serverUrl;
     }
 
+    @Deprecated
+    @Restricted(DoNotUse.class)
+    @RestrictedSince("2.2.0")
+    @CheckForNull
+    @NonNull
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMNavigator.java]: Does not make sense that I can both be `@CheckForNull` and `@NonNull`

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMNavigator.java [hunk 14]
```diff
@@ -359,6 +362,18 @@ public String getBitbucketServerUrl() {
         return serverUrl;
     }
 
+    @Deprecated
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMNavigator.java]: Should not be marked deprecated as it is actively used.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/AbstractBitbucketEndpoint.java [hunk 15]
```diff
@@ -83,6 +120,218 @@
...
+        }
+
+        // The non-null not-empty bitbucketJenkinsRootUrl after the update
+        // above is an already processed and normalized string
+        LOGGER.log(Level.FINEST, "AbstractBitbucketEndpoint::getEndpointJenkinsRootUrl : original: {0}", "'" + endpointcfgEJRU + "'" );
+        return endpointcfgEJRU;
+    }
+
+    @Restricted(NoExternalUse.class)
+    @Deprecated
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/AbstractBitbucketEndpoint.java]: Should not be marked `@Deprecated`

**Comment by jimklimov** [comment]
I will try to wrap my head around the second set of changes, I did not quite understand the difference, so will try to look for existing samples to replicate :)

How required are these "minor improvements" to accept the PR? Is it feasible e.g. for you to merge "as is" and apply the needed "proper Java" fixes faster and more correctly than I can? :)

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/BitbucketCloudEndpoint.java (hunk 9 above)
Comment (jimklimov) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/BitbucketCloudEndpoint.java]: Why shouldn't that field be added to the constructor? It is a common property of all descendants of the abstract class, including our Cloud and Server configs, so empty or not - it is sort of required and always set... But maybe I just misunderstand the magic behind this instantiation.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/AbstractBitbucketEndpoint.java (hunk 7 above)
Comment (jimklimov) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/AbstractBitbucketEndpoint.java]: My understanding was that `final` is used for configuration values, so effectively set once per object lifetime (a reconfig creates a new object with same `serverUrl` for consumers to reference, and removes the current one), and the keyword enforces that?

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/BitbucketCloudEndpoint.java (hunk 9 above)
Comment (jimklimov) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/BitbucketCloudEndpoint.java]: Also, looking now at where to make the change, I remembered that this abstract class did not provide any setters for configurable values, so I was not sure it is possible/proper to add any. Instead, all the common variables (credentials and Manage boolean) were set via constructor.

**Comment by casz** [comment]
> How much "required" are these "minor improvements" to accept the PR? Is it feasible e.g. for you to merge "as is" and apply the needed "proper Java" fixes faster and more correctly than I can? :)

For the databound constructor I would say they are hard requirements, based on experience and jenkins dev docs.
See https://jenkins.io/doc/developer/plugin-development/pipeline-integration/

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/AbstractBitbucketEndpoint.java (hunk 7 above)
Comment (casz) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/AbstractBitbucketEndpoint.java]: it should not be marked final since it can be changed after the fact when the config is created.

**Comment by casz** [comment]
If you would like I can attempt to implement the requested stuff :)

**Comment by jimklimov** [comment]
Yes, if you would, please :) Seems like only the cherry on top is now missing?

**Comment by casz** [comment]
@jimklimov what was your reasoning for not normalizing the URL when the user saves configuration?

**Comment by casz** [comment]
Fyi, instead of ugly logging, you can attach a debugger to Jenkins.

With `mvn hpi:run` you get a Jenkins instance spun up for you. Inside IntelliJ or NetBeans it as simple as creating a task/run configuration and hit the "run debug" button and you are golden.

**Comment by jimklimov** [comment]
Principle of least surprise i guess, when user saves something (or edits
xml's directly) might be surprised to see a different string end up there.
Though after some evolution, we do use the different string in practice so
point may be moot, and this double-accounting can be slashed too. Note to
not store the default global root url if setting from user is empty, and to
reevaluate it every time, though - it can change "live" and unknown to this
class (or add more caching vars?) as end-user reconfigures stuff.

**Comment by casz** [comment]
I think it would be best to drop the calculated value it adds too much complexity.

The simpler would be to rely on `bitbucketJenkinsRootUrl` and then if it is null use the Jenkins root url

**Comment by casz** [comment]
I think this is in a mergeable state, the question is does it solve your problem @jimklimov ? :sweat_smile:

**Comment by jimklimov** [comment]
Well, I'll have to test on our server to be sure, hope to feed back soon.
Maybe Monday, maybe before.

Also to be sure about concerns I had about processing the global root url,
I'd like to see some tests that reconfigure it (to different quality values
WRT normalization and slash) and check that a non-customized endpoint
returns the good spelling of current root url. That would make me confident
about the merge :-)

**Comment by casz** [comment]
> Also to be sure about concerns I had about processing the global root url, I'd like to see some tests that reconfigure it (to different quality values WRT normalization and slash) and check that a non-customized endpoint returns the good spelling of current root url. That would make me confident about the merge :-)

I would not bother, Jenkins admin user is free to change the URL so :man_shrugging:

**Comment by jimklimov** [comment]
> I would not bother, Jenkins admin user is free to change the URL so man_shrugging

Well, the idea here is that if the user does not fill out the custom URL, which is the default state, then the current global value, whatever it is each time a Root URL is requested for webhook configuration, should be used. So changing the global config should apply to webhook runs that happen after that change.

**Comment by jimklimov** [comment]
Ok, got no strict position on e.g. normalizing away port numbers, trailing slash enforcement should suffice in practice. I'm adding the test to document the expected behavior and make sure it (or similar) sanity stays in place, and IMHO can merge :)

**Comment by jimklimov** [comment]
PS: Tested that it still posts just the expected hook URLs in practice :)
